README.md
[![Build Status](https://travis-ci.org/nycmouli/PYDFT.svg?branch=master)](https://travis-ci.org/nycmouli/PYDFT)
[![Coverage Status](https://coveralls.io/repos/github/nycmouli/PYDFT/badge.svg?branch=master)](https://coveralls.io/github/nycmouli/PYDFT?branch=master)

# PYDFT
My personal DFT code
 PYDFT- Python Density Funtion Theory code.




History.md
# Revision History

## Revision 0.0.1

- Poisson solver is completed. Code written as pre the method described in Tomas Arias lectures.
- Unit tests are written for each funtion in poisson solver.


tests/test_poisson.py
import numpy as np
import pytest
from dftCode.fourierBasis import Ope, Lope, LinvOpe, cI, cJ
from dftCode.geometry import _M, _N, gen_r_G_G2
from dftCode.charge_dis import charge_dis
from dftCode.poisson import phi



def test_phi():
    """
     Tests the oneline poisson solver funtion.
    
    Poisson solver in a single line of code."""
    
    assert np.allclose(phi(charge_dis([2,2,2],np.diag([2,2,2]),[0.75,0.5],[1,-1]),[2,2,2],np.diag([2,2,2])),np.array([ 0.11648337+0.j,  0.07467416+0.j,  0.07467416+0.j, -0.01777448+0.j,
    0.07467416+0.j, -0.01777448+0.j, -0.01777448+0.j, -0.28718243+0.j]))


tests/test_geometry.py
import numpy as np
import pytest
from dftCode.geometry import _M, _N, gen_r_G_G2
import sys


def test_M():
    """Tests that the _M subroutine works.                                                                                                                   
    """
    assert _M([2,2,2])==[[0, 0, 0],[0, 0, 1], [0, 1, 0], [0, 1, 1], [1, 0, 0], [1, 0, 1], [1, 1, 0], [1, 1, 1]]
    assert _M([3,2,1])==[[0, 0, 0], [0, 1, 0], [1, 0, 0], [1, 1, 0], [2, 0, 0], [2, 1, 0]]


def test_N():
    """Tests that the  _N subroutine works.
    This is another way to test arrays. using np.allclose
    """
    assert np.allclose(_N([3,2,1]),np.array([[ 0,  0,  0],[ 0,  1,  0],[ 1,  0,  0],[ 1,  1,  0],[-1,  0,  0],[-1,  1,  0]])) == True
    assert np.allclose(_N([2,2,2]),np.array([[0, 0, 0],
                                             [0, 0, 1],
                                             [0, 1, 0],
                                             [0, 1, 1],
                                             [1, 0, 0],
                                             [1, 0, 1],
                                             [1, 1, 0],
                                             [1, 1, 1]]))

def test_gen_r_G_G2():
    """ Tests weather gen_r_G_G2 subroutine works
    """
    assert np.allclose(gen_r_G_G2([2,2,2],np.diag([2,2,2]),'r'),np.array([[ 0.,  0.,  0.],
                                                                           [ 0.,  0.,  1.],
                                                                           [ 0.,  1.,  0.],
                                                                           [ 0.,  1.,  1.],
                                                                           [ 1.,  0.,  0.],
                                                                           [ 1.,  0.,  1.],
                                                                           [ 1.,  1.,  0.],
                                                                           [ 1.,  1.,  1.]]))
    assert np.allclose(gen_r_G_G2([2,2,2],np.diag([2,2,2]),'G'),np.array([[ 0.        ,  0.        ,  0.        ],
                                                                           [ 0.        ,  0.        ,  3.14159265],
                                                                           [ 0.        ,  3.14159265,  0.        ],
                                                                           [ 0.        ,  3.14159265,  3.14159265],
                                                                           [ 3.14159265,  0.        ,  0.        ],
                                                                           [ 3.14159265,  0.        ,  3.14159265],
                                                                           [ 3.14159265,  3.14159265,  0.        ],
                                                                           [ 3.14159265,  3.14159265,  3.14159265]])) 
    assert np.allclose(gen_r_G_G2([2,2,2],np.diag([2,2,2]),'G2'),np.array([  0.       ,   9.8696044,   9.8696044,  19.7392088,   9.8696044,
                                                                              19.7392088,  19.7392088,  29.6088132]))


tests/test_charge_dis.py
import numpy as np
import pytest
from dftCode.geometry import _M, _N, gen_r_G_G2
from dftCode.charge_dis import gaussian,charge_dis


def test_gaussian():
    """ Tests the gaussian charge distribution routine in charge_dis.py
    """
    assert np.allclose(gaussian([0.75,0.5],[1,-1],np.linspace(0,1,10)),np.array([-0.35744565, -0.34669986, -0.31613924, -0.27038447, -0.21590651,
       -0.15959883, -0.10743921, -0.06358218, -0.03003532, -0.00686962])) == True
    assert np.allclose(gaussian([0.75,0.5,0.3],[1,-1,0.2],np.linspace(0,1,10)),np.array([ 0.11287757,  0.09244664,  0.04133802, -0.01668859, -0.05894094,
       -0.07493042, -0.06762259, -0.04725793, -0.02420049, -0.00505139]))

def test_charge_dis():
    """ Tests the charge_dis subroutine in charge_dis.py                                                                                                 
    """
    assert np.allclose(charge_dis([2,2,2],np.diag([2,2,2]),[0.75,0.5],[1,-1]),np.array([ 0.00919842,  0.01613367,  0.01613367, -0.00686962,  0.01613367,
      -0.00686962, -0.00686962, -0.35744565])) == True






dftCode/base.py
import warnings
def deprecated(func):
    '''This is a decorator which can be used to mark functions
    as deprecated. It will result in a warning being emitted
    when the function is used.'''
    def new_func(*args, **kwargs):
        warnings.warn("Call to deprecated function {}.".format(func.__name__),
                      category=DeprecationWarning)
        return func(*args, **kwargs)
    new_func.__name__ = func.__name__
    new_func.__doc__ = func.__doc__
    new_func.__dict__.update(func.__dict__)
    return new_func

def exhandler(function, parser):
    """If -examples was specified in 'args', the specified function
    is called and the application exits.

    :arg function: the function that prints the examples.
    :arg parser: the initialized instance of the parser that has the
      additional, script-specific parameters.
    """
#    print("Hello---base")
    args = vars(bparser.parse_known_args()[0])
#    print(args)
    if args["examples"]:
        function()
        return
    if args["verbose"]:
        from liveserial.msg import set_verbosity
        set_verbosity(args["verbose"])

    args.update(vars(parser.parse_known_args()[0]))
    return args

def _common_parser():
    """Returns a parser with common command-line options for all the scripts
    in the fortpy suite.
    """
    import argparse
    parser = argparse.ArgumentParser(add_help=False)
    parser.add_argument("-examples", action="store_true",
                        help="See detailed help and examples for this script.")
    parser.add_argument("-verbose", action="store_true",
                        help="See verbose output as the script runs.")
    parser.add_argument('-action', nargs=1, choices=['save','print'], default='print',
                        help="Specify what to do with the output (print or save)")
    parser.add_argument("-debug", action="store_true",
                        help="Print verbose calculation information for debugging.")

    return parser

bparser = _common_parser()
testmode = False
"""bool: when True, the package is operating in unit test mode, which changes
how plotting is handled.
"""
def set_testmode(testing):
    """Sets the package testing mode.
    """
    global testmode
    testmode = testing


dftCode/geometry.py

import numpy as np
# For performing linear algebra operations using matrices
from numpy import linalg as LA
from numpy.linalg import inv
import operator
# for Generating tuples list.
import itertools 


def _M(s):
    """Defining the 'M' matrix.
    'M' has coordinates in matrix form for each sample point in real space.
    We use M to get list of real space sample  points 'r'
    r=M(Diag(S))^(-1)R^(Transpose)
    
     Args:
        input:
            s => number of sample points along the lattice vectors.
        output: 
            M matrix. 
    """
    M = tuple(itertools.product(*[range(0,s[0]),range(0,s[1]),range(0,s[2])]))
    M=[list(i) for i in M]
    return M



def _N(s):
    """Defining the 'N' matrix.
    'N' has coordinates in matrix form for each sample point in reciprocal
    space.
    We use N to get list of reciprocal lattice vectos 'G'
    G= 2*pi*N*R^(inverse)
    
    Args:
        input:
            s => number of sample points along the lattice vectors.
        output: 
            N matrix. 
    """
    M=_M(s)
    m1=np.transpose(M)[0]
    m2=np.transpose(M)[1]
    m3=np.transpose(M)[2]
    n1=[ m1[i]-s[0] if m1[i] > s[0]/2 else m1[i] for i in range(len(m1)) ]
    n2=[ m2[i]-s[1] if m2[i] > s[1]/2 else m2[i] for i in range(len(m2)) ]
    n3=[ m3[i]-s[2] if m3[i] > s[2]/2 else m3[i] for i in range(len(m3)) ]
    N=np.transpose([n1,n2,n3])
    return N

def gen_r_G_G2(s,R,p=None):
    """Generates r, G and G2 matrices.
    
    Args: input:
           s => number of sample points along the lattice vectors. 
           R=[a,b,c] => Defines the lattice vectors. (a,b,c) are any +ve integers.
           p = 'r', 'G' or 'G2'(square of G matrix) 
           
          Output:
           p='r' => Outputs r matrix.
           p='G' => Outputs G matrix.
           p='2' => Outputs G2 matrix.
       
       Example: s = [15,15,15]
                R= np.diag([6,6,6])
                _gen_r_G_G2(s,R,"r")
    """
    if p is None:
        return "Please pass a third argument asking for 'r', 'G' or 'G2'."
    if p is "r":
        r=np.dot(np.dot(_M(s),inv(np.diag(s))),np.transpose(R))
        return r
    if p is "G":
        G=2*np.pi*np.dot(_N(s),inv(R))
        return G
    if p is "G2":
        G=2*np.pi*np.dot(_N(s),inv(R))
        G2=np.sum(np.transpose(G*G),axis=0) 
        return G2
    else:
        return "Error: Passed an argument other than r, G or G2."


dftCode/charge_dis.py
import numpy as np

from dftCode.geometry import gen_r_G_G2


def gaussian(sig,coeff,x):
    """ Defining Gaussian charge distribution at point x in 3D space.
    Args:
        inputs: 
            sig => sigmas for each Gaussian.
            coeff => coefficients for each gaussian funtion.
                Note: All the gaussians will be summed up.
            x (np.array()): List of input points 
                Note: x in 3D space is x=Sqrt(a^2+b^2+c^2) where
                {a,b,c} are the cartesian coordinates.
            
        Output: Charge distribution at points x. 
        
    """
    g=lambda x,sigma: (np.exp(-(x**2)/(2*sigma**2)))/(2*np.pi*sigma**2)**(3./2)
    n=lambda x: np.sum([coeff[i]*g(x,sig[i]) for i in range(len(sig))],axis=0)
    if len(sig)!=len(coeff):
        return "Error: Length of 'sig' not equal to length of 'coeff'."
    else:
        return n(x)


def charge_dis(s,R,sig,coeff):
    """Generates the charge distribution at each sample point for the unit cell.
        We change the midpoint of the unitcell to origin here.
    
    Args:
        input:
            R=[a,b,c] => Defines the lattice vectors. (a,b,c) are any +ve integers.
            s=[a,b,c] => Defines the number of sample points along the
                        lattice vectors.  (a,b,c) are any integers.
    """
    # There is a problem with sum()
    mid = [sum(R[i])/2. for i in range(len(R))] #sum(R,axis=1)/2.
    #Note: Need to check if this mid is right for non diagnol R vectors.
    #dr=np.sqrt(sum((gen_r_G_G2(s,R,"r")-mid)**2,axis=1))
    ss=(gen_r_G_G2(s,R,"r")-mid)**2
    dr=np.sqrt([sum(ss[i]) for i in range(len(ss))])
    return gaussian(sig,coeff,dr)
        


dftCode/fourierBasis.py
"""Plane wave basis for the DFT solvers. Exposes the generic linear
algebra functions required by the other solvers and methods.
"""


# For performing linear algebra operations using matrices
from numpy import linalg as LA
from numpy.linalg import inv
import operator
# for Generating tuples list.
import itertools 


from dftCode import msg
from dftCode.geometry import gen_r_G_G2

import numpy as np


def Ope(inp,R):
    """Outputs the overlap matrix. In our case here we are specific to plane wave basis.
    Args:
        input:
            inp (numpy array): The overlap matrix operator acts on this input.
            R=[a,b,c] => Defines the lattice vectors. (a,b,c) are any +ve integers.
    """
    return np.linalg.det(R)*np.dot(np.identity(len(inp)),inp)   


def Lope(inp,s,R):
    """Lope is the L operator as defined in Tomas arias Lecture.
     Note:Works only for plane wave basis.
    Args:
        input:
            inp (numpy array): The overlap matrix operator acts on this input.
            s=[a,b,c] => Defines the number of sample points along the
                        lattice vectors.  (a,b,c) are any integers.
            R=[a,b,c] => Defines the lattice vectors. (a,b,c) are any +ve integers.
    """
    if len(gen_r_G_G2(s,R,"G2")) == len(inp):
        return -np.linalg.det(R)*np.dot(np.diag(gen_r_G_G2(s,R,"G2")),inp)
    else:
        return "Error: Input vector size is not same as G2"

def LinvOpe(inp,s,R):
    """LinvOpe is the inverse of L operator as defined in Tomas arias Lecture.
     Note:Works only for plane wave basis.
    Args:
        input:
            inp (numpy array): The overlap matrix operator acts on this input.
            s=[a,b,c] => Defines the number of sample points along the
                        lattice vectors.  (a,b,c) are any integers.
            R=[a,b,c] => Defines the lattice vectors. (a,b,c) are any +ve integers.
    """
    if len(gen_r_G_G2(s,R,"G2")) == len(inp):
        Lele=-np.linalg.det(R)*np.diag(gen_r_G_G2(s,R,"G2"))
        Lele[0][0]=1
        Leleinv=inv(Lele)
        Leleinv[0][0]=0
        return np.dot(Leleinv,inp)
    else:
        return "Error: Input vector size is not same as G2"

def cI(inp,s,R,p=None):
    """cI is the fourier series expansion coeff Matrix. It is the forward transform matrix for the problem.
        Note:Works only for plane wave basis.
        Args:
        input:
            inp (numpy array): The overlap matrix operator acts on this input.
            s=[a,b,c] => Defines the number of sample points along the
                        lattice vectors.  (a,b,c) are any integers.
            R=[a,b,c] => Defines the lattice vectors. (a,b,c) are any +ve integers.
            p (default): Fast fourier transform. #Written with Wiley's code help.
                p='s' => slow fourier transform.
    """
    if p is None:
        f=np.fft.fftn(inp.reshape(s,order="F")).reshape(np.prod(s),order="F")
    elif p is "s": 
        f=np.dot(np.exp(1j*np.dot(gen_r_G_G2(s,R,"G"),np.transpose(gen_r_G_G2(s,R,"r")))),inp)
    return f

def cJ(inp,s,R,p=None):
    """cI is the fourier series expansion coeff Matrix. You can get the fast fourier transform too.
      Note:Works only for plane wave basis.
    Args:
        input:
            inp (numpy array): The overlap matrix operator acts on this input.
            s=[a,b,c] => Defines the number of sample points along the
                        lattice vectors.  (a,b,c) are any integers.
            R=[a,b,c] => Defines the lattice vectors. (a,b,c) are any +ve integers.
    """
    if p is None:
        f=np.fft.ifftn(inp.reshape(s,order="F")).reshape(np.prod(s),order="F")
        return f
    elif p is "s":
        f=np.exp(1j*np.dot(gen_r_G_G2(s,R,"G"),np.transpose(gen_r_G_G2(s,R,"r"))))
        return (1./np.prod(s))*np.dot(np.transpose(f.conjugate()),inp)




dftCode/dftSolve.py
#!/usr/bin/python

from dftCode import msg
import numpy as np
import csv


def examples():
    """Prints examples of using the script to the console using colored output.                                                   
    """
    script = "Code to compute DFT energies of a crystal structure."
    explain = ("This code produces a numerical solution for the energy of a system of "
               "atoms. As of now we have only the poisson solver.")
    contents = [(("Find the charge density for a simple cubic cell with lattice parameter of 6 "
                  "with 6 divisions along each lattice vector."),
                 "dftSolve.py 6 -s [6,6,6]",
                 "This prints the value of potential evaluated at all the sample points "
                 "to the screen")]
    required = ("REQUIRED: The lattice parameter to be specified `a`.")
    output = ("RETURNS: A potential file named `potential.csv` if -poisson is specified.")
    details = (".")
    outputfmt = ("")

    msg.example(script, explain, contents, required, output, outputfmt, details)


# script options is for arg parse. I'm not intentding to use it. But I dont want to delete it now.

script_options = {
    "-la": dict(default=1., type=float,
              help=("Lattice parameter for the crystal structure.")),
    "-crystal": dict(default="sc", type=str,
                     help=("The type of primitive cell to use options are (sc)")),
    "-s": dict(default=[6,6,6], nargs="+",
              help=("Number of sampling points along each basis vector.")),
    "-poisson": dict(action="store_true",
                     help=("Prints the solution to the poisson equation to the screen.")),
    
     }

# _parser_options is also for arg parse. I'm not intending to use it. I will use config parse.

def _parser_options():
    """Parses the options and arguments from the command line."""
    #We have two options: get some of the details from the config file,                                                  
    import argparse
    from dftCode import base

    pdescr = "Numerical DFT code."
    parser = argparse.ArgumentParser(parents=[base.bparser], description=pdescr)
    for arg, options in script_options.items():
        parser.add_argument(arg, **options)

    

    args = base.exhandler(examples, parser)

    if args is None:
        return

    return args # pragma: no cover 


# The run below runs with args. I wrote a better config parse code below. So not intented to use this function.
def run(args):
    """Runs the code to find the energy of the crystal using DFT.
    """

    if args is None:
        print "No args given"
        exit

    if args["poisson"]:

        from dftCode.fourierBasis import Ope, Lope, LinvOpe, cJ, cI
        from dftCode.poisson import phi
        from dftCode.charge_dis import charge_dis
        
        R = args["la"]*np.array([[1,0,0],[0,1,0],[0,0,1]])
        s = [int(i) for i in args["s"]]
        
        sig=[0.75,0.5]
        coeff=[-1,1]

        # Analytic energy for Poisson
        Uanalytic=((1/sig[0] + 1/sig[1])/2 -np.sqrt(2)/np.sqrt(sig[0]**2+sig[1]**2))/np.sqrt(np.pi)

        # Numerical energy solved using 1 line poisson solver.

        Unumerical=0.5*np.real(np.dot(np.conjugate(cJ(np.real(phi(charge_dis(s,R,sig,coeff),s,R)),s,R)),Ope(cJ(charge_dis(s,R,sig,coeff),s,R),R)))

        
        print "Uanalytic =", Uanalytic
        print"Unumerical =" , Unumerical

        return 0



def ConfigSectionMap(section):

    # The code has to work for anyone who downloads.                                                                                                                      
    # So, DATA_PATH gives the path to config file.                                                                                                                       
    import os
    this_dir, this_filename = os.path.split(__file__)
    tmp=this_dir.split("/")
    actual_path="/".join(tmp[:-1])
    DATA_PATH = os.path.join(actual_path, "InpDFT", "input_dft.cfg")



    # Writing config parser part with help from https://wiki.python.org/moin/ConfigParserExamples                                                                         
    import ConfigParser
    Config = ConfigParser.ConfigParser()
    Config.read(DATA_PATH)

    dict1 = {}
    options = Config.options(section)
    for option in options:
        try:
            dict1[option] = Config.get(section, option)
            if dict1[option] == -1:
                DebugPrint("skip: %s" % option)
        except:
            print("exception on %s!" % option)
            dict1[option] = None
    return dict1


def checkFloat(i):
    try:
        float(i)
        return True
    except:
        print("Warning: Check your input file. 'Spaces' or any other 'wrong input' are not yet handled.")
        return False


def run_config():
    """Runs the code using "input_dft.cfg" to find the energy of the crystal using DFT    
    """


    from dftCode.fourierBasis import Ope, Lope, LinvOpe, cJ, cI
    from dftCode.poisson import phi
    from dftCode.charge_dis import charge_dis


    geo = ConfigSectionMap("geometry")['geo']
    rIn = ConfigSectionMap("geometry")['r']
    sIn = ConfigSectionMap("geometry")['s']
    
    if(geo!="simpleCubic"):
        print("Only simple Cubic geometry is programmed now. \n Please pass in geo=simpleCubic in config file.")
    else:

        tmp=rIn.split(" ")
        R=[int(i) for i in tmp if checkFloat(i)] 
        R = R*np.array([[1,0,0],[0,1,0],[0,0,1]])
        #print(R)

        tmp=sIn.split(" ")
        s=[int(i) for i in tmp if checkFloat(i)]
        #print(s)

    
    char_dis = ConfigSectionMap("chargeDis")['char_dis']
    sigIn = ConfigSectionMap("chargeDis")['sig']
    coeffIn=ConfigSectionMap("chargeDis")['coeff']


    if(char_dis!="gaussian"):
        print("Only Gaussian distribution is programmed now. \n Please pass in 'char_dis=gaussian' in config file.")
    else:

        tmp=sigIn.split(" ")
        sig=[float(i) for i in tmp if checkFloat(i)]
        #print(sig)

        tmp=coeffIn.split(" ")
        coeff=[int(i) for i in tmp if checkFloat(i)]
        #print(coeff)

        # Analytic energy for Poisson                                                                                                                                      
        Uanalytic=((1/sig[0] + 1/sig[1])/2 -np.sqrt(2)/np.sqrt(sig[0]**2+sig[1]**2))/np.sqrt(np.pi)

        # Numerical energy solved using 1 line poisson solver.                                                                                                           
        Unumerical=0.5*np.real(np.dot(np.conjugate(cJ(np.real(phi(charge_dis(s,R,sig,coeff),s,R)),s,R)),Ope(cJ(charge_dis(s,R,sig,coeff),s,R),R)))

        print "Uanalytic =", Uanalytic
        print"Unumerical =" , Unumerical
        
    return 0

     
if __name__ == '__main__': # pragma: no cover
    #run(_parser_options()) # As of now I do not want to run with arg parse.
    run_config() # Runs with config file.
